File: quadrantos/Sources/QuadrantOS/ScreenDBOverlay.swift
//
//  ScreenDBOverlay.swift
//  QuadrantOS
//
//  Transparent fullscreen overlay — a cross dividing the screen into 4 quadrants.
//  Each quadrant shows live ScreenDB data overlaid on top of any desktop window.
//
//  Top-Left:     ◇ Windows + ⟡ Accessibility
//  Top-Right:    ⌁ Processes
//  Bottom-Left:  ◆ Receipts + Ledger
//  Bottom-Right: ◉ ScreenDB Status
//
//  Data comes from native Swift APIs — no Python Process calls:
//  - AppleScript (osascript) for windows and accessibility tree
//  - ps for processes
//  - SQLite for receipts and ledger
//

import SwiftUI
import AppKit
import Foundation
import SQLite3

// MARK: - Overlay Window Controller

final class ScreenDBOverlayController {
    static let shared = ScreenDBOverlayController()
    private var window: NSPanel?

    func show() {
        guard window == nil else { return }
        guard let mainScreen = NSScreen.main else { return }
        let screen = mainScreen.frame
        let panel = NSPanel(
            contentRect: screen,
            styleMask: [.borderless, .nonactivatingPanel],
            backing: .buffered,
            defer: false
        )
        panel.level = .statusBar
        panel.collectionBehavior = [.canJoinAllSpaces, .stationary, .ignoresCycle]
        panel.isOpaque = false
        panel.backgroundColor = .clear
        panel.hasShadow = false
        panel.isMovableByWindowBackground = false
        panel.hidesOnDeactivate = false
        panel.contentView = NSHostingView(rootView: ScreenDBOverlayView())
        window = panel
        panel.orderFrontRegardless()
    }

    func hide() {
        window?.orderOut(nil)
        window = nil
    }

    func toggle() {
        if window != nil { hide() } else { show() }
    }
}

// MARK: - Live Data Feed

final class ScreenDBLiveFeed: ObservableObject {
    @Published var windows: [WindowInfo] = []
    @Published var elements: [ElementInfo] = []
    @Published var processes: [ProcessInfo] = []
    @Published var receipts: [ReceiptInfo] = []
    @Published var ledgerValid: Bool = true
    @Published var ledgerCount: Int = 0
    @Published var frontmostApp: String = ""
    @Published var frontmostTitle: String = ""
    @Published var timestamp: String = ""
    @Published var tick: Int = 0

    private var timer: Timer?

    struct WindowInfo: Identifiable, Hashable {
        let id: String; let app: String; let title: String; let focused: Bool
        let x: Int; let y: Int; let w: Int; let h: Int
    }
    struct ElementInfo: Identifiable, Hashable {
        let id: String; let role: String; let title: String; let value: String
        let enabled: Bool; let focused: Bool; let x: Int; let y: Int
    }
    struct ProcessInfo: Identifiable, Hashable {
        let id: String; let pid: Int; let name: String; let cpu: Double
    }
    struct ReceiptInfo: Identifiable, Hashable {
        let id: String; let success: Bool; let type: String; let output: String
        let shotBefore: String; let shotAfter: String; let changed: Bool
    }

    func start() {
        stop()
        refresh()
        timer = Timer.scheduledTimer(withTimeInterval: 2.0, repeats: true) { _ in self.refresh() }
    }

    func stop() { timer?.invalidate(); timer = nil }

    private func refresh() {
        tick += 1
        timestamp = DateFormatter.localizedString(from: Date(), dateStyle: .none, timeStyle: .medium)
        DispatchQueue.global(qos: .utility).async {
            let wins = self.captureWindows()
            let els = self.captureElements()
            let procs = self.captureProcesses()
            let (recs, count) = self.readReceipts()
            DispatchQueue.main.async {
                self.windows = wins.0; self.frontmostApp = wins.1; self.frontmostTitle = wins.2
                self.elements = els; self.processes = procs
                self.receipts = recs; self.ledgerCount = count
            }
        }
    }

    private func runOSA(_ script: String) -> String {
        let task = Process()
        task.launchPath = "/usr/bin/osascript"
        task.arguments = ["-e", script]
        let pipe = Pipe()
        task.standardOutput = pipe
        task.standardError = Pipe()
        do {
            try task.run(); task.waitUntilExit()
            return String(data: pipe.fileHandleForReading.readDataToEndOfFile(), encoding: .utf8) ?? ""
        } catch { return "" }
    }

    private func captureWindows() -> ([WindowInfo], String, String) {
        let script = """
        tell application "System Events"
            set output to ""
            set frontApp to ""
            set frontWin to ""
            repeat with proc in (every process whose background only is false)
                try
                    set procName to name of proc
                    set isFront to (proc is frontmost)
                    if isFront then
                        set frontApp to procName
                        try
                            set frontWin to name of front window of proc
                        end try
                    end if
                    repeat with w in (every window of proc)
                        try
                            set wTitle to name of w
                            set wPos to position of w
                            set wSize to size of w
                            set output to output & procName & "|" & wTitle & "|" & (item 1 of wPos) & "|" & (item 2 of wPos) & "|" & (item 1 of wSize) & "|" & (item 2 of wSize) & "|" & isFront & "\\n"
                        end try
                    end repeat
                end try
            end repeat
            return output & "FRONT|" & frontApp & "|" & frontWin
        end tell
        """
        let output = runOSA(script)
        var wins: [WindowInfo] = []; var frontApp = ""; var frontWin = ""
        for line in output.components(separatedBy: "\n") {
            if line.hasPrefix("FRONT|") {
                let p = line.components(separatedBy: "|")
                if p.count >= 3 { frontApp = p[1]; frontWin = p[2] }
                continue
            }
            let p = line.components(separatedBy: "|")
            if p.count >= 7 {
                wins.append(WindowInfo(id: "w\(wins.count)", app: p[0], title: p[1],
                    focused: p[6] == "true", x: Int(p[2]) ?? 0, y: Int(p[3]) ?? 0,
                    w: Int(p[4]) ?? 0, h: Int(p[5]) ?? 0))
            }
        }
        return (wins, frontApp, frontWin)
    }

    private func captureElements() -> [ElementInfo] {
        let script = """
        tell application "System Events"
            set output to ""
            set frontProc to first process whose frontmost is true
            try
                repeat with el in (every UI element of front window of frontProc)
                    try
                        set elRole to role of el
                        set elTitle to ""
                        try
                            set elTitle to title of el
                        end try
                        set elValue to ""
                        try
                            set elValue to value of el
                        end try
                        set elPos to {0, 0}
                        try
                            set elPos to position of el
                        end try
                        set elEnabled to true
                        try
                            set elEnabled to enabled of el
                        end try
                        set output to output & elRole & "|" & elTitle & "|" & elValue & "|" & (item 1 of elPos) & "|" & (item 2 of elPos) & "|" & elEnabled & "\\n"
                    end try
                end repeat
            end try
            return output
        end tell
        """
        let output = runOSA(script)
        var els: [ElementInfo] = []
        for line in output.components(separatedBy: "\n") {
            let p = line.components(separatedBy: "|")
            if p.count >= 6 {
                let title = p[1] == "missing value" ? "" : p[1]
                let value = p[2] == "missing value" ? "" : p[2]
                els.append(ElementInfo(id: "e\(els.count)", role: p[0],
                    title: String(title.prefix(20)), value: String(value.prefix(15)),
                    enabled: p[5] == "true", focused: false,
                    x: Int(p[3]) ?? 0, y: Int(p[4]) ?? 0))
            }
        }
        return els
    }

    private func captureProcesses() -> [ProcessInfo] {
        let task = Process()
        task.launchPath = "/bin/ps"
        task.arguments = ["-A", "-o", "pid=,comm=,%cpu=", "-r"]
        let pipe = Pipe()
        task.standardOutput = pipe; task.standardError = Pipe()
        var procs: [ProcessInfo] = []
        do {
            try task.run(); task.waitUntilExit()
            let output = String(data: pipe.fileHandleForReading.readDataToEndOfFile(), encoding: .utf8) ?? ""
            for line in output.components(separatedBy: "\n").prefix(10) {
                let parts = line.trimmingCharacters(in: .whitespaces).components(separatedBy: .whitespaces).filter { !$0.isEmpty }
                if parts.count >= 3 {
                    procs.append(ProcessInfo(id: "p\(procs.count)", pid: Int(parts[0]) ?? 0,
                        name: String((parts[1] as NSString).lastPathComponent.prefix(20)),
                        cpu: Double(parts[2]) ?? 0))
                }
            }
        } catch {}
        return Array(procs.prefix(8))
    }

    private func readReceipts() -> ([ReceiptInfo], Int) {
        let dbPath = NSHomeDirectory() + "/Library/Application Support/ScreenDB/screendb.db"
        var db: OpaquePointer?
        var receipts: [ReceiptInfo] = []; var count = 0
        guard sqlite3_open(dbPath, &db) == SQLITE_OK else { sqlite3_close(db); return ([], 0) }
        var stmt: OpaquePointer?
        if sqlite3_prepare_v2(db, "SELECT receipt_id, action_type, success, result, screenshot_before_hash, screenshot_after_hash FROM actions ORDER BY ts DESC LIMIT 6", -1, &stmt, nil) == SQLITE_OK {
            while sqlite3_step(stmt) == SQLITE_ROW {
                let rid = String(cString: sqlite3_column_text(stmt, 0))
                let type = String(cString: sqlite3_column_text(stmt, 1))
                let success = sqlite3_column_int(stmt, 2) == 1
                let result = String(cString: sqlite3_column_text(stmt, 3))
                let shotB = String(cString: sqlite3_column_text(stmt, 4))
                let shotA = String(cString: sqlite3_column_text(stmt, 5))
                receipts.append(ReceiptInfo(id: "r\(receipts.count)", success: success, type: type,
                    output: String(result.prefix(40)), shotBefore: String(shotB.prefix(6)),
                    shotAfter: String(shotA.prefix(6)), changed: shotB != shotA))
            }
            sqlite3_finalize(stmt)
        }
        var cstmt: OpaquePointer?
        if sqlite3_prepare_v2(db, "SELECT COUNT(*) FROM actions", -1, &cstmt, nil) == SQLITE_OK {
            if sqlite3_step(cstmt) == SQLITE_ROW { count = Int(sqlite3_column_int(cstmt, 0)) }
            sqlite3_finalize(cstmt)
        }
        sqlite3_close(db)
        return (receipts, count)
    }
}

// MARK: - Overlay View

struct ScreenDBOverlayView: View {
    @StateObject private var feed = ScreenDBLiveFeed()

    var body: some View {
        ZStack {
            Color.black.opacity(0.12).ignoresSafeArea()
            GeometryReader { geo in
                let cx = geo.size.width / 2
                let cy = geo.size.height / 2
                ZStack {
                    // The Cross
                    Rectangle().fill(Color.orange.opacity(0.6)).frame(width: 2).position(x: cx, y: cy)
                    Rectangle().fill(Color.orange.opacity(0.6)).frame(height: 2).position(x: cx, y: cy)
                    // Crosshair
                    ZStack {
                        Circle().stroke(Color.orange.opacity(0.7), lineWidth: 1.5).frame(width: 30, height: 30)
                        Circle().fill(Color.orange.opacity(0.8)).frame(width: 4, height: 4)
                    }.position(x: cx, y: cy)

                    // Q0: Top-Left — Windows + Accessibility
                    QuadrantContent(title: "◇ WINDOWS", subtitle: "⟡ ACCESSIBILITY", glyph: "◧", color: .cyan) {
                        WindowsQuadrant(feed: feed)
                    }
                    .frame(width: cx - 8, height: cy - 8)
                    .position(x: cx / 2, y: cy / 2)

                    // Q1: Top-Right — Processes
                    QuadrantContent(title: "⌁ PROCESSES", subtitle: "top 8 by CPU", glyph: "◨", color: .green) {
                        ProcessesQuadrant(feed: feed)
                    }
                    .frame(width: cx - 8, height: cy - 8)
                    .position(x: cx + cx / 2, y: cy / 2)

                    // Q2: Bottom-Left — Receipts
                    QuadrantContent(title: "◆ RECEIPTS", subtitle: feed.ledgerValid ? "✓ INTACT \(feed.ledgerCount)" : "✗ BROKEN", glyph: "◪", color: .orange) {
                        ReceiptsQuadrant(feed: feed)
                    }
                    .frame(width: cx - 8, height: cy - 8)
                    .position(x: cx / 2, y: cy + cy / 2)

                    // Q3: Bottom-Right — Status
                    QuadrantContent(title: "◉ SCREENDB", subtitle: "tick #\(feed.tick)", glyph: "◨", color: .purple) {
                        StatusQuadrant(feed: feed)
                    }
                    .frame(width: cx - 8, height: cy - 8)
                    .position(x: cx + cx / 2, y: cy + cy / 2)
                }
            }
        }
        .onAppear { feed.start() }
        .onDisappear { feed.stop() }
    }
}

// MARK: - Quadrant Container

struct QuadrantContent<Content: View>: View {
    let title: String; let subtitle: String; let glyph: String; let color: Color
    @ViewBuilder let content: () -> Content
    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack(spacing: 6) {
                Text(glyph).font(.system(size: 10, design: .monospaced)).foregroundColor(color.opacity(0.5))
                Text(title).font(.system(size: 10, weight: .bold, design: .monospaced)).foregroundColor(color.opacity(0.8))
                Spacer()
                Text(subtitle).font(.system(size: 8, design: .monospaced)).foregroundColor(.gray.opacity(0.5))
            }
            .padding(.horizontal, 10).padding(.vertical, 4)
            ScrollView { content().padding(.horizontal, 10).padding(.vertical, 4) }
        }
        .background(color.opacity(0.03))
        .overlay(RoundedRectangle(cornerRadius: 0).stroke(color.opacity(0.12), lineWidth: 1))
    }
}

// MARK: - Quadrant Views

struct WindowsQuadrant: View {
    @ObservedObject var feed: ScreenDBLiveFeed
    var body: some View {
        VStack(alignment: .leading, spacing: 2) {
            if !feed.frontmostApp.isEmpty {
                HStack(spacing: 4) {
                    Text("◉").font(.system(size: 9, design: .monospaced)).foregroundColor(.orange)
                    Text(feed.frontmostApp).font(.system(size: 10, weight: .bold, design: .monospaced)).foregroundColor(.orange)
                    Text("| \(feed.frontmostTitle)").font(.system(size: 8, design: .monospaced)).foregroundColor(.gray).lineLimit(1)
                }.padding(.bottom, 4)
            }
            ForEach(feed.windows) { w in
                HStack(spacing: 6) {
                    Text(w.focused ? "◉" : "◌").font(.system(size: 9, design: .monospaced)).foregroundColor(w.focused ? .orange : .gray.opacity(0.4))
                    Text(w.app).font(.system(size: 9, design: .monospaced)).foregroundColor(w.focused ? .white : .gray).frame(width: 90, alignment: .leading)
                    Text(w.title).font(.system(size: 8, design: .monospaced)).foregroundColor(.gray.opacity(0.6)).lineLimit(1)
                    Spacer()
                    Text("\(w.w)×\(w.h)").font(.system(size: 7, design: .monospaced)).foregroundColor(.gray.opacity(0.4))
                }
            }
            if !feed.elements.isEmpty {
                Divider().background(Color.gray.opacity(0.1)).padding(.vertical, 4)
                Text("⟡ ELEMENTS").font(.system(size: 8, weight: .bold, design: .monospaced)).foregroundColor(.cyan.opacity(0.5)).padding(.bottom, 2)
            }
            ForEach(feed.elements) { e in
                HStack(spacing: 4) {
                    Text(e.enabled ? "✓" : "✗").font(.system(size: 8, design: .monospaced)).foregroundColor(e.enabled ? .green : .red)
                    Text(e.role).font(.system(size: 8, design: .monospaced)).foregroundColor(.cyan.opacity(0.7)).frame(width: 80, alignment: .leading)
                    Text(e.title).font(.system(size: 8, design: .monospaced)).foregroundColor(.gray.opacity(0.6)).lineLimit(1)
                    if !e.value.isEmpty {
                        Text("v=\(e.value)").font(.system(size: 7, design: .monospaced)).foregroundColor(.purple.opacity(0.5))
                    }
                }
            }
        }
    }
}

struct ProcessesQuadrant: View {
    @ObservedObject var feed: ScreenDBLiveFeed
    var body: some View {
        VStack(alignment: .leading, spacing: 2) {
            ForEach(feed.processes) { p in
                HStack(spacing: 6) {
                    Text("\(p.pid)").font(.system(size: 8, design: .monospaced)).foregroundColor(.gray.opacity(0.4)).frame(width: 50, alignment: .leading)
                    Text(p.name).font(.system(size: 9, design: .monospaced)).foregroundColor(.gray).frame(width: 100, alignment: .leading)
                    GeometryReader { geo in
                        ZStack(alignment: .leading) {
                            RoundedRectangle(cornerRadius: 1).fill(Color.gray.opacity(0.1)).frame(height: 3)
                            RoundedRectangle(cornerRadius: 1).fill(LinearGradient(colors: [.green, .orange, .red], startPoint: .leading, endPoint: .trailing))
                                .frame(width: max(2, min(geo.size.width, geo.size.width * min(1.0, p.cpu / 50))), height: 3)
                        }
                    }.frame(height: 3)
                    Text(String(format: "%.1f%%", p.cpu)).font(.system(size: 8, design: .monospaced)).foregroundColor(.orange).frame(width: 45, alignment: .trailing)
                }
            }
        }
    }
}

struct ReceiptsQuadrant: View {
    @ObservedObject var feed: ScreenDBLiveFeed
    var body: some View {
        VStack(alignment: .leading, spacing: 2) {
            HStack {
                Text(feed.ledgerValid ? "✓ LEDGER INTACT" : "✗ LEDGER BROKEN").font(.system(size: 9, weight: .bold, design: .monospaced)).foregroundColor(feed.ledgerValid ? .green : .red)
                Spacer()
                Text("\(feed.ledgerCount) total").font(.system(size: 8, design: .monospaced)).foregroundColor(.gray.opacity(0.5))
            }.padding(.bottom, 4)
            if feed.receipts.isEmpty {
                Text("◌ No receipts yet").font(.system(size: 9, design: .monospaced)).foregroundColor(.gray.opacity(0.4))
            }
            ForEach(feed.receipts) { r in
                HStack(spacing: 4) {
                    Text(r.success ? "✓" : "✗").font(.system(size: 8, design: .monospaced)).foregroundColor(r.success ? .green : .red)
                    Text(r.id).font(.system(size: 8, design: .monospaced)).foregroundColor(.gray.opacity(0.5))
                    Text(r.type).font(.system(size: 8, design: .monospaced)).foregroundColor(.cyan.opacity(0.7)).frame(width: 70, alignment: .leading)
                    Text(r.output).font(.system(size: 7, design: .monospaced)).foregroundColor(.gray.opacity(0.5)).lineLimit(1)
                    Spacer()
                    Text("\(r.shotBefore)→\(r.shotAfter)").font(.system(size: 7, design: .monospaced)).foregroundColor(.gray.opacity(0.4))
                    if r.changed { Text("⟁").font(.system(size: 8, design: .monospaced)).foregroundColor(.orange) }
                }
            }
        }
    }
}

struct StatusQuadrant: View {
    @ObservedObject var feed: ScreenDBLiveFeed
    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Text("◉ LIVE").font(.system(size: 12, weight: .bold, design: .monospaced)).foregroundColor(.green)
                Spacer()
                Text(feed.timestamp).font(.system(size: 8, design: .monospaced)).foregroundColor(.gray.opacity(0.4))
            }
            Divider().background(Color.gray.opacity(0.1))
            HStack(spacing: 16) {
                StatBlock(label: "WINDOWS", value: "\(feed.windows.count)", color: .cyan)
                StatBlock(label: "ELEMENTS", value: "\(feed.elements.count)", color: .cyan)
                StatBlock(label: "PROCS", value: "\(feed.processes.count)", color: .green)
                StatBlock(label: "RECEIPTS", value: "\(feed.ledgerCount)", color: .orange)
            }
            Divider().background(Color.gray.opacity(0.1))
            VStack(alignment: .leading, spacing: 3) {
                Text("ARCHITECTURE").font(.system(size: 8, weight: .bold, design: .monospaced)).foregroundColor(.gray.opacity(0.5))
                Text("ScreenCaptureKit sees").font(.system(size: 8, design: .monospaced)).foregroundColor(.gray.opacity(0.6))
                Text("AXUIElement understands").font(.system(size: 8, design: .monospaced)).foregroundColor(.gray.opacity(0.6))
                Text("CursorAgent acts").font(.system(size: 8, design: .monospaced)).foregroundColor(.gray.opacity(0.6))
                Text("ReceiptOS proves").font(.system(size: 8, design: .monospaced)).foregroundColor(.gray.opacity(0.6))
            }
            Spacer()
            HStack { Spacer(); Text("tick #\(feed.tick)").font(.system(size: 7, design: .monospaced)).foregroundColor(.gray.opacity(0.3)) }
        }
    }
}

struct StatBlock: View {
    let label: String; let value: String; let color: Color
    var body: some View {
        VStack {
            Text(value).font(.system(size: 16, weight: .bold, design: .monospaced)).foregroundColor(color)
            Text(label).font(.system(size: 7, design: .monospaced)).foregroundColor(.gray.opacity(0.5))
        }
    }
}

File: screendb_overlay.py
#!/usr/bin/env python3
"""
ScreenDB Overlay — Transparent fullscreen cross dividing screen into 4 quadrants.
Each quadrant shows live ScreenDB data overlaid on top of any desktop window.

Uses tkinter with transparent background on macOS.
"""
import sys, os, json, time, subprocess, threading, tkinter as tk
from datetime import datetime

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from screendb import capture_accessibility, get_receipts, verify_receipts
from mac_runtime import capture_windows, capture_processes


class ScreenDBOverlay:
    def __init__(self):
        self.root = tk.Tk()
        self.root.title("ScreenDB Overlay")
        self.root.attributes("-fullscreen", True)
        self.root.attributes("-topmost", True)
        self.root.configure(bg="black")
        self.root.attributes("-alpha", 0.85)

        self.screen_w = self.root.winfo_screenwidth()
        self.screen_h = self.root.winfo_screenheight()
        self.cx = self.screen_w // 2
        self.cy = self.screen_h // 2

        self.canvas = tk.Canvas(self.root, bg="black", highlightthickness=0,
                                width=self.screen_w, height=self.screen_h)
        self.canvas.pack(fill="both", expand=True)

        self.running = True
        self.data = {
            "windows": [], "elements": [], "processes": [],
            "receipts": [], "frontmost": "", "frontmost_title": "",
            "ledger_valid": True, "ledger_count": 0, "tick": 0,
        }

        # Colors
        self.ORANGE = "#ff8c00"
        self.CYAN = "#00ddff"
        self.GREEN = "#00ff88"
        self.RED = "#ff3344"
        self.PURPLE = "#aa44ff"
        self.GRAY = "#606070"
        self.WHITE = "#e0e0e8"
        self.DIM = "#303040"

        # Bind Escape to quit
        self.root.bind("<Escape>", lambda e: self.quit())
        self.root.bind("<Button-1>", lambda e: None)  # click-through attempt

        # Start data thread
        self.thread = threading.Thread(target=self._data_loop, daemon=True)
        self.thread.start()

        # Start render loop
        self._render()
        self.root.mainloop()

    def quit(self):
        self.running = False
        self.root.destroy()

    def _data_loop(self):
        while self.running:
            try:
                windows = capture_windows()
                ax = capture_accessibility()
                procs = capture_processes()
                receipts = get_receipts(6)
                v = verify_receipts()

                self.data = {
                    "windows": windows[:10],
                    "elements": ax.get("elements", [])[:12],
                    "processes": sorted(procs, key=lambda p: p.get("cpu", 0), reverse=True)[:8],
                    "receipts": receipts,
                    "frontmost": ax.get("frontmost_app", ""),
                    "frontmost_title": ax.get("frontmost_window", ""),
                    "ledger_valid": v["intact"],
                    "ledger_count": v["total"],
                    "tick": self.data["tick"] + 1,
                    "timestamp": datetime.now().strftime("%H:%M:%S"),
                }
            except Exception:
                pass
            time.sleep(2)

    def _render(self):
        if not self.running:
            return
        self.canvas.delete("all")
        d = self.data
        cx, cy = self.cx, self.cy

        # ─── The Cross ──────────────────────────────
        # Vertical line
        self.canvas.create_line(cx, 0, cx, self.screen_h, fill=self.ORANGE, width=2, stipple="")
        # Horizontal line
        self.canvas.create_line(0, cy, self.screen_w, cy, fill=self.ORANGE, width=2)

        # Crosshair circle
        self.canvas.create_oval(cx - 15, cy - 15, cx + 15, cy + 15, outline=self.ORANGE, width=1)
        self.canvas.create_oval(cx - 2, cy - 2, cx + 2, cy + 2, fill=self.ORANGE, outline="")

        # ─── Q0: Top-Left — Windows + Accessibility ─
        qx0, qy0 = 4, 4
        qw, qh = cx - 8, cy - 8
        self._draw_quadrant_header(qx0, qy0, "◇ WINDOWS  ⟡ ACCESSIBILITY", self.CYAN)
        y = qy0 + 25
        if d["frontmost"]:
            self.canvas.create_text(qx0 + 8, y, text=f"◉ {d['frontmost']} | {d['frontmost_title'][:35]}",
                                    fill=self.ORANGE, font=("SF Mono", 11, "bold"), anchor="w")
            y += 18
        for w in d["windows"][:6]:
            mark = "◉" if w["focused"] else "◌"
            color = self.WHITE if w["focused"] else self.GRAY
            self.canvas.create_text(qx0 + 8, y, text=f"  {mark} {w['app'][:18]:<18} {w['title'][:30]}",
                                    fill=color, font=("SF Mono", 10), anchor="w")
            y += 15
        if d["elements"]:
            y += 5
            self.canvas.create_text(qx0 + 8, y, text="⟡ ELEMENTS", fill=self.CYAN, font=("SF Mono", 9, "bold"), anchor="w")
            y += 14
            for e in d["elements"][:6]:
                en = "✓" if e.get("enabled") else "✗"
                en_color = self.GREEN if e.get("enabled") else self.RED
                title = e.get("title", "") or ""
                val = e.get("value", "") or ""
                val_s = f' v="{val[:15]}"' if val and val != "missing value" else ""
                self.canvas.create_text(qx0 + 8, y, text=f"  {en} {e['role'][:16]:<16} {title[:20]}{val_s}",
                                        fill=self.GRAY, font=("SF Mono", 9), anchor="w")
                y += 13

        # ─── Q1: Top-Right — Processes ──────────────
        qx1 = cx + 4
        self._draw_quadrant_header(qx1, qy0, "⌁ PROCESSES  top 8 by CPU", self.GREEN)
        y = qy0 + 25
        for p in d["processes"][:8]:
            name = p["name"][:20]
            cpu = p.get("cpu", 0)
            bar_len = int(min(40, cpu * 2))
            bar = "█" * bar_len
            self.canvas.create_text(qx1 + 8, y, text=f"  {p['pid']:>7} {name:<20} {cpu:>5.1f}% {bar}",
                                    fill=self.GRAY, font=("SF Mono", 9), anchor="w")
            y += 14

        # ─── Q2: Bottom-Left — Receipts ─────────────
        qy2 = cy + 4
        ledger_text = f"✓ INTACT {d['ledger_count']}" if d["ledger_valid"] else "✗ BROKEN"
        ledger_color = self.GREEN if d["ledger_valid"] else self.RED
        self._draw_quadrant_header(qx0, qy2, "◆ RECEIPTS", self.ORANGE, badge=ledger_text, badge_color=ledger_color)
        y = qy2 + 25
        if not d["receipts"]:
            self.canvas.create_text(qx0 + 8, y, text="◌ No receipts yet — run: python3 screendb.py exec notification --target test --value hello",
                                    fill=self.DIM, font=("SF Mono", 8), anchor="w")
            y += 14
        for r in d["receipts"][:6]:
            ok = "✓" if r["result"]["success"] else "✗"
            ok_color = self.GREEN if r["result"]["success"] else self.RED
            changed = "⟁" if r["screenshot"]["before_hash"] != r["screenshot"]["after_hash"] else " "
            shot_b = r["screenshot"]["before_hash"][:6]
            shot_a = r["screenshot"]["after_hash"][:6]
            atype = r["action"]["type"][:14]
            output = r["result"]["output"][:35]

            self.canvas.create_text(qx0 + 8, y, text=f"  {ok} [{r['receipt_id'][:12]}] {atype:>14} {output}",
                                    fill=self.GRAY, font=("SF Mono", 9), anchor="w")
            self.canvas.create_text(qx0 + 8 + 600, y, text=f"{shot_b}→{shot_a} {changed}",
                                    fill=self.ORANGE if changed.strip() else self.DIM, font=("SF Mono", 8), anchor="w")
            y += 14

        # ─── Q3: Bottom-Right — Status ──────────────
        self._draw_quadrant_header(qx1, qy2, "◉ SCREENDB", self.PURPLE, badge=f"tick #{d['tick']}", badge_color=self.GRAY)
        y = qy2 + 25
        ts = d.get("timestamp", "")
        self.canvas.create_text(qx1 + 8, y, text=f"  ◉ LIVE  {ts}", fill=self.GREEN, font=("SF Mono", 12, "bold"), anchor="w")
        y += 22

        self.canvas.create_text(qx1 + 8, y, text=f"  WINDOWS   {len(d['windows'])}", fill=self.CYAN, font=("SF Mono", 14, "bold"), anchor="w")
        self.canvas.create_text(qx1 + 120, y, text=f"  ELEMENTS  {len(d['elements'])}", fill=self.CYAN, font=("SF Mono", 14, "bold"), anchor="w")
        y += 20
        self.canvas.create_text(qx1 + 8, y, text=f"  PROCS     {len(d['processes'])}", fill=self.GREEN, font=("SF Mono", 14, "bold"), anchor="w")
        self.canvas.create_text(qx1 + 120, y, text=f"  RECEIPTS  {d['ledger_count']}", fill=self.ORANGE, font=("SF Mono", 14, "bold"), anchor="w")
        y += 28

        self.canvas.create_text(qx1 + 8, y, text="  ARCHITECTURE", fill=self.DIM, font=("SF Mono", 9, "bold"), anchor="w")
        y += 14
        for line in ["  ScreenCaptureKit sees", "  AXUIElement understands", "  CursorAgent acts", "  ReceiptOS proves"]:
            self.canvas.create_text(qx1 + 8, y, text=line, fill=self.GRAY, font=("SF Mono", 9), anchor="w")
            y += 13

        # ─── Footer ─────────────────────────────────
        self.canvas.create_text(self.screen_w - 10, self.screen_h - 10,
                                text="ESC to quit · tick #%d · %s" % (d["tick"], ts),
                                fill=self.DIM, font=("SF Mono", 8), anchor="se")

        self.root.after(500, self._render)

    def _draw_quadrant_header(self, x, y, title, color, badge="", badge_color=None):
        self.canvas.create_text(x + 8, y, text=title, fill=color, font=("SF Mono", 10, "bold"), anchor="w")
        if badge:
            bc = badge_color or self.GRAY
            self.canvas.create_text(x + self.cx - 20, y, text=badge, fill=bc, font=("SF Mono", 8), anchor="e")
        # Border line under header
        self.canvas.create_line(x, y + 14, x + self.cx - 12, y + 14, fill=self.DIM, width=1)


if __name__ == "__main__":
    overlay = ScreenDBOverlay()
